### ISSUE TITLE
Dynamic data type array

### ISSUE THREAD

**Issue by crash-g** [rust-ndarray/ndarray#651]
Hi,
I'm trying to write something like the following:
```rust
fn create_dynamic_data_type_array(is_float: bool) -> ??? {
    if is_float {
        ArrayD::<f64>::zeros(IxDyn(&[10, 10]))
    } else {
        ArrayD::<u64>::zeros(IxDyn(&[10, 10]))
    }
}
```
**USE CASE:**
I'd like to write a microservice that consumes Numpy arrays serialized using Protobuf and does some basic manipulation over them (slicing and similar). `rust-ndarray` looks like a great candidate, but I'm not sure how to abstract over the data type short of defining a custom `enum` with the relevant variants and manually forwarding methods of interest.

Is there an easy way to do this (I'm sort of new to Rust and I may be missing something)? Or otherwise is this feature out of scope for this library?
Thanks!

**Comment by jturner314** [comment]
Here are a few options. I recommend option 1 because it's straightforward, has the best performance, and is simple to implement. If that doesn't work, IMO option 2 is the next-best option, but it's a lot more tedious.

1. Write the implementation generically over the element type, and dispatch between types at the outermost level of your service, with something like this:

   ```rust
   use ndarray::prelude::*;
   use num_traits::Num;
   use std::fmt::{Debug, Display};

   fn process_array<A>(a: ArrayD<A>)
   where
       // Specify whatever trait bounds you need.
       A: Num + Clone + Debug + Display,
   {
       // The implementation is generic over the element type.
       println!("{}", a);
   }

   fn main() {
       let is_float = false;
       // Dispatch over the element type at the outermost level of your service.
       // The implementation of `process_array` is generic over the element type.
       if is_float {
           process_array(ArrayD::<f64>::zeros(IxDyn(&[10, 10])));
       } else {
           process_array(ArrayD::<i64>::zeros(IxDyn(&[10, 10])));
       }
   }
   ```

2. Wrap arrays in an enum and then forward methods on that enum to the inner arrays. For example, you might have

   ```rust
   pub enum NumArcArrayD {
       Float(ArcArray<f64, IxDyn>),
       Integer(ArcArray<i64, IxDyn>),
   }
   ```

   This isn't too bad; it just might be tedious to write methods forwarding everything you need. This would be roughly equivalent to how [NumPy is implemented](https://docs.scipy.org/doc/numpy/reference/c-api.types-and-structures.html#c.PyArrayObject).

3. Similar to option 2, but use a [trait object](https://doc.rust-lang.org/book/ch17-02-trait-objects.html) (`Box<YourArrayTrait>`) instead of an enum. You'd need to write your own trait `YourArrayTrait` that would be implemented by all the array types you're working with.

4. Wrap each element in an enum individually. This would be expensive, and it wouldn't make sure that all the elements are the same variant within the array.

5. Similar to option 4, but use a trait object (`Box<YourElementTrait>`) as the element type instead of an enum. This would be expensive because each element would be individually heap-allocated.

6. Use dynamic typing with [`std::any`](https://doc.rust-lang.org/std/any/index.html), e.g.

   ```rust
   use std::any::{Any, TypeId};

   fn do_whatever(a: &dyn Any) {
       let type_id = a.type_id();
       if type_id == TypeId::of::<ArrayD<f64>>() {
           let array = a.downcast_ref::<ArrayD<f64>>().unwrap();
           println!("{}", array);
       } else if type_id == TypeId::of::<ArrayD<i64>>() {
           let array = a.downcast_ref::<ArrayD<i64>>().unwrap();
           println!("{}", array);
       } else {
           unimplemented!()
       }
   }
   ```

   and you could write a macro to eliminate the repetitiveness. I generally recommend avoiding `Any`, though.

7. Convert all array inputs to `f64`. :)

Does option 1 work for your use-case?

**Comment by crash-g** [comment]
Thanks a lot for the comprehensive list of options!
Option 2 was what I had in mind but option 1 looks better. Either one will do for my use case.
I was not aware of option 6, good to know it exists though it is definitely not pretty.

Out of curiosity, option 2 could be provided by the library in theory: is there a specific reason why you'd rather not?

**Comment by jturner314** [comment]
> I was not aware of option 6, good to know it exists though it is definitely not pretty.

I realized that the example with `Any` could be written a little nicer:

```rust
fn do_whatever(a: &dyn Any) {
    if let Some(array) = a.downcast_ref::<ArrayD<f64>>() {
        println!("{}", array);
    } else if let Some(array) = a.downcast_ref::<ArrayD<i64>>() {
        println!("{}", array);
    } else {
        unimplemented!()
    }
}
```

I still don't recommend that approach, though.

> Out of curiosity, option 2 could be provided by the library in theory: is there a specific reason why you'd rather not?

`ndarray::ArrayBase` is generic over the element type (the element type can be anything). `ndarray` tries to be as flexible as possible. In contrast, an enum has a fixed number of variants. NumPy works around this issue by treating Python objects as one of the possible data types (`np.object_`). We could do something similar:

```rust
// TODO: Should there be any other supertraits? Should `Any` really be included?
trait Number: Any + Num + Clone + Debug + Display {}

impl<T> Number for T where T: Any + Num + Clone + Debug + Display {}

pub enum NumArcArray<D: Dimension> {
    F32(ArcArray<f32, D>),
    F64(ArcArray<f64, D>),
    I64(ArcArray<i64, D>),
    U64(ArcArray<u64, D>),
    // TODO: Add 8-bit, 16-bit, 32-bit, 128-bit integer variants
    // TODO: Add bool?
    C32(ArcArray<Complex32, D>),
    C64(ArcArray<Complex64, D>),
    Any(ArcArray<Box<dyn Number>, D>),
}
```

where the `Any` variant is analogous to NumPy's handling of Python objects. This generally works, but it has a number of disadvantages:

1. For any type that doesn't have its own variant (even a thin wrapper around a numeric type), the `Any` variant must be used, which comes with a very large performance cost.

2. It's unclear what the trait bounds should be. Are those that I included sufficient? Are they too restrictive?

3. It's unclear what types should have their own variants. I've included the standard numeric types plus complex numbers, but what about other useful types, like `num::Rational`?

I don't think this design is suitable for inclusion in `ndarray`. It's selecting a few "special" types which will have good performance, but everything else will have very bad performance. It's not clear what variants and trait bounds should be included.

I think this type of approach could be reasonable for user code where the user knows exactly what set of types to include in the enum. It doesn't work so well for the `ndarray` crate, where we'd have to decide what to include to try to satisfy everyone.

If we wanted to offer functionality like this, I think it would make the most sense to provide a macro that would automatically implement methods on the user's specific enum type. (IMO, this would make sense in crate separate from the core `ndarray` crate, at least initially.)
